syntax = "proto3";

package users;

option go_package = "./";

import "google/protobuf/timestamp.proto";

message User {
  string id = 1;
  string name = 2;
  int64 telegram_id = 3;
}

message Tutor {
  User user = 1;
  repeated string tags = 2;
}

message TutorDetails {
  Tutor tutor = 1;
  string bio = 2;
  int32 response_count = 3; // это только для своего профиля
  repeated Review review = 4;
  repeated string tags = 5;
  bool is_active = 6; // это только для своего профиля
  google.protobuf.Timestamp created_at = 7;
}

message Review {
  string id = 1;
  string tutor_id = 2;
  string student_id = 3;
  int32 rating = 4;
  string comment = 5;
  google.protobuf.Timestamp created_at = 6;
}

service UserService {
  rpc CreateUser(CreateUserRequest) returns (CreateResponse);
  rpc AddResponsesToTutor(AddResponseToTutorRequest) returns (AddResponseToTutorResponse);
  rpc CreateNewResponse(CreateResponseRequest) returns (Success);
  rpc GetUserById(GetById) returns (User);
  rpc GetTutorById(GetById) returns (Tutor);
  rpc GetTutorInfoById(GetById) returns (TutorDetails);
  rpc GetStudentById(GetById) returns (User);
  rpc GetUserByTelegramId(GetByTelegramId) returns (User);
  rpc GetAllUsers(GetAllRequest) returns (GetAllResponse);
  rpc GetAllTutorsPagination(GetAllTutorsPaginationRequest) returns (GetTutorsPaginationResponse);
  rpc UpdateBioTutor(UpdateBioRequest) returns (UpdateBioResponse);
  rpc UpdateTags(UpdateTagsRequest) returns (Success);
  rpc ChangeTutorActive(SetActiveTutorById) returns (Success);
  rpc ChangeTutorName(ChangeNameRequest) returns (Success);
  rpc CreateReview(CreateReviewRequest) returns (CreateReviewResponse);
  rpc GetReview(GetReviewRequest) returns (Review);
  rpc GetReviews(GetReviewsRequest) returns (GetReviewsResponse);
}

message GetAllTutorsPaginationRequest {
  int32 page = 1;
  int32 size = 2;
}

message UpdateBioRequest {
  string id = 1;
  string bio = 2;
}

message ChangeNameRequest {
  string id = 1;
  string name = 2;
}

message UpdateBioResponse {
  bool success = 1;
}

message CreateUserRequest {
  int64 telegram_id = 1;
  string name = 2;
  string role = 3;
}

message CreateResponse {
  string id = 1;
}

message CreateResponseRequest {
  string tutor_id = 1;
}

message AddResponseToTutorRequest {
  int64 tutor_id = 1;
  int32 response_count = 2;
}

message AddResponseToTutorResponse {
  int32 response_count = 1;
  bool success = 2;
}

message GetByTelegramId {
  int64 id = 1;
}

message SetActiveTutorById {
  string id = 1;
  bool active = 2;
}

message GetById {
  string id = 1;
}

message GetAllRequest {}

message GetAllResponse {
  repeated Tutor tutors = 1;
}

message GetTutorsPaginationResponse {
  int64 count = 1;
  repeated Tutor tutors = 2;
}

message UpdateTagsRequest {
  string tutor_id = 1;
  repeated string tags = 2;
}

message Success {
  bool success = 1;
}

message CreateReviewRequest {
  string tutor_id = 1;
  string student_id = 2;
  int32 rating = 3;
  string comment = 4;
}

message CreateReviewResponse {
  string id = 1;
}

message GetReviewRequest {
  string id = 1;
}

message GetReviewsRequest {
  string tutor_id = 1;
}

message GetReviewsResponse {
  repeated Review reviews = 1;
}
